nixos/modules/services/misc/duckling.nix

```diff
@@ -0,0 +1,64 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.duckling;
+
+  port = cfg.port;
+in {
+  options = {
+    services.duckling = {
+      enable = mkEnableOption "duckling";
+
+      dataDir = mkOption {
+        type = types.path;
+        default = "/var/lib/duckling";
+        description = ''
+          Directory to store duckling's logs.
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.haskellPackages.duckling;
+        defaultText = "pkgs.haskellPackages.duckling";
+        example = literalExample "pkgs.haskellPackages.duckling";
+        description = "Version of duckling to run";
+      };
+
+      port = mkOption {
+        type = types.int;
+        default = 8080;
+        description = ''
+          Port on which duckling will run.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.users.duckling = {
```

Comment:
I'm guessing we don't need to create a dedicated user for this service and `DynamicUser` would be sufficient. Are you familiar with `DynamicUser`?

Reply:
No, I'm not. Do you have any pointers?

Reply:
[Documentation](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#DynamicUser=) has a reasonable explanation. In this service you can simply add `serviceConfig.DynamicUser = true;` and I think it will work.

Reply:
Worth noting is that dynamic users can only write to `StateDirectory`, `LogsDirectory`, and `CacheDirectory` paths - all of which must exist under `/var/lib`. You might want to remove the `dataDir` option in this case, letting `systemd` take care of it via `serviceConfig.LogsDirectory = "duckling";`.

Reply:
Changed to a DynamicUser